src/Components/test/E2ETest/Tests/FormsTest.cs

```diff
@@ -230,8 +227,9 @@ public void InputDateInteractsWithEditContext_NullableDateTimeOffset()
             Browser.Equal("modified valid", () => expiryDateInput.GetAttribute("class"));
 
             // Can become invalid
-            expiryDateInput.Clear();
-            expiryDateInput.SendKeys("111111111");
+            // Note that we have to update it via JS not via SendKeys, because date inputs are special. Selenium's mechanism for
+            // simulating keystrokes causes date inputs to behave inconsistently in this specific scenario.
+            ChangeDateInputField(".expiry-date input", "11111-11-11");
```

Comment:
I didn't run into this problem in the [globalization tests](https://github.com/aspnet/AspNetCore/blob/master/src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs#L245). did you try combining 'clear' and sending keys?

Reply:
I'm happy with whatever approach we take here, but it would be nice if we are consistent across tests. Is this something that would benefit from a helper that we could use everywhere? I would prefer if we move to a mode where we create small reusable solutions for these problems instead of dealing directly with Selenium, as that way we avoid tripping into these issues.

Reply:
> I didn't run into this problem in the globalization tests

The other forms tests with date inputs don't run into it either. The special thing about this particular test is that it's a nullable datetime, which is really hard to make invalid, because almost any sequence of keystrokes into an `<input type=date>` will get coerced into some kind of valid datetime or null. The `11111-11-11` case is basically the only data entry type I've found that doesn't get coerced to something valid or null.

The globalization tests don't try to cause validation failures AFAICT.

> did you try combining 'clear' and sending keys?

The code before already did call `Clear` before sending keys.

> I'm happy with whatever approach we take here, but it would be nice if we are consistent across tests.

I tend to prefer siding with the most "native" way to interact with the browser that we can, since it's most likely to represent actual human-UI interaction. Using JS to mess with the DOM directly is more of a last resort than something good to spread out throughout the E2E tests, since maybe we stop representing scenarios that actually can occur in reality. It's annoying we have to do it here but I don't know of a better alternative.

Alternatively we could decide that this one test is really artificial since it's so unlikely a human would produce this invalid-entry case and just remove it. But there is code we have that does something useful with this case so typically we do want to have an E2E test.

Reply:
> The code before already did call Clear before sending keys.

Right, what I meant was combing them into a single edit operation (like the code I liked). This means doing a select-all (ctrl+a) and then sending the keys.

Reply:
OK, thanks for clarifying. Yes - that's a good idea. I have no way to know if it works 100% in this scenario without actually trying it, but it definitely seems worth trying. We will have to see what pass rate we get in practice.

I've updated this PR to use that technique, and also to factor out the `ReplaceText` code into a reusable extension method on `IWebElement` since it's not as artificial as mutating the DOM directly.